src/ride_type.cpp

```diff
@@ -687,10 +692,13 @@ void RidesManager::DeleteInstance(uint16 num)
 	this->instances[num]->RemoveAllPeople();
 	_guests.NotifyRideDeletion(this->instances[num]);
 
-	for (const XYZPoint16& position : this->instances[num]->get_all_positions()) {
-		Voxel& voxel = *_world.GetCreateVoxel(position, false);
-		assert(voxel.instance == num + SRI_FULL_RIDES);
-		voxel.ClearInstances();
+	for (const XYZPoint16& position : this->instances[num]->GetAllPiecePositions()) {
+		Voxel *voxel = _world.GetCreateVoxel(position, false);
+		assert(voxel);
+		if (voxel->instance != SRI_FREE) {
+			assert(voxel->instance == num + SRI_FULL_RIDES);
+			voxel->ClearInstances();
+		}
```

Comment:
A `RidesManager` deals with a ride at global level, high level stuff. How muich money does it make etc.
For some reason, you decided that it must now deal with trackpiece positions, something internal to the ride. I don't know why you did that.

In a real world comparison, what you're doing is giving the CEO of an amusement park a list of track pieces he should remove. Th CEO must climb into the ride, and remove each piece by him/her self.

So what you should do instead is follow the chain of command here. A CEO tells the ride owner to remove the ride, the ride owner tells the work people to remove all track pierces, and the work people figure out which voxels they have the change.
So here, you'd just say `this->instances[num]->RemoveFromWorld();`  this is what a CEO can handle comfortably.
Inside the Ride instance, you then pass it on to all parts of the ride, which then actually change voxels.

In that way, you keep internals at the level where they belong, and you're not duplicating code to other parts of the code base that eventually will get inconsistent.